File: labs/mobile/donor-archive/tunnel/NodeIngestDeduplicator.kt
package com.luminet.android.tunnel

data class AndroidNodeEndpointKey(
    val host: String,
    val port: Int,
    val protocol: String
)

data class AndroidScrapedNode(
    val host: String,
    val port: Int,
    val protocol: String,
    val source: String,
    val pingMs: Long,
    val isAlive: Boolean
)

class NodeIngestDeduplicator {
    private val seenEndpoints = mutableSetOf<AndroidNodeEndpointKey>()
    private val uniqueNodes = mutableListOf<AndroidScrapedNode>()
    private val sourceStats = mutableMapOf<String, Int>()

    fun ingestNode(node: AndroidScrapedNode): Boolean {
        val key = AndroidNodeEndpointKey(
            host = node.host.trim().lowercase(),
            port = node.port,
            protocol = node.protocol.trim().lowercase()
        )

        return if (seenEndpoints.add(key)) {
            sourceStats[node.source] = (sourceStats[node.source] ?: 0) + 1
            uniqueNodes.add(node)
            true
        } else {
            false
        }
    }

    fun getRankedNodes(): List<AndroidScrapedNode> {
        return uniqueNodes
            .filter { it.isAlive }
            .sortedBy { it.pingMs }
    }

    fun totalUnique(): Int = uniqueNodes.size

    fun countForSource(source: String): Int = sourceStats[source] ?: 0
}

File: labs/control-ui/api-shadow/node_ingest_deduplicator.ts
export interface ScrapedNodeInfo {
  host: string;
  port: number;
  protocol: string;
  source: string;
  pingMs: number;
  isAlive: boolean;
}

export class NodeIngestDeduplicator {
  private seenEndpoints = new Set<string>();
  private uniqueNodes: ScrapedNodeInfo[] = [];
  private sourceStats = new Map<string, number>();

  ingestNode(node: ScrapedNodeInfo): boolean {
    const key = `${node.host.trim().toLowerCase()}:${node.port}:${node.protocol.trim().toLowerCase()}`;
    if (this.seenEndpoints.has(key)) {
      return false;
    }
    this.seenEndpoints.add(key);
    this.sourceStats.set(node.source, (this.sourceStats.get(node.source) || 0) + 1);
    this.uniqueNodes.push(node);
    return true;
  }

  getRankedNodes(): ScrapedNodeInfo[] {
    return this.uniqueNodes
      .filter((n) => n.isAlive)
      .sort((a, b) => a.pingMs - b.pingMs);
  }

  totalUnique(): number {
    return this.uniqueNodes.length;
  }

  countForSource(source: string): number {
    return this.sourceStats.get(source) || 0;
  }
}
